### ISSUE TITLE
Inconsistent handling of POST request body across different media types

### ISSUE THREAD

**Issue by lisandromc** [spring-projects/spring-boot#9349]
Request params fail to be consistently extracted from request body when using different request content-types.

The `params` property of the `@PostMapping` annotation seems to be the trigger to param capture failure.

With WWW forms there is a workaround: capturing the body through a `@RequestParam` instead of `@RequestBody`. That workaround is no good with JSON, though.

This happens at least with JSON and WWW URL-encoded form and Spring Boot version 1.5.3.RELEASE:

This seems like a bug to me. Or am I getting something wrong here?

```java
@Controller
public class TestController {

  // body is retrieved fine
  @PostMapping(path="/json",
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostBody(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }

  // 400 Bad Request, no params are retrieved therefore required params aren't met
  @PostMapping(path="/json_with_params", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostBodyAndParams(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }

  // 415 Unsupported media type
  // Content type 'application/x-www-form-urlencoded;charset=UTF-8' not supported
  @PostMapping(path="/form_with_params", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
  public ResponseEntity<String> processFormPostBodyAndParams(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }

  // body is retrieved fine, but only as long as it's injected as @RequestParam, not as @RequestBody
  @PostMapping(path="/form_with_params_skip_body", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
  public ResponseEntity<String> processFormPostAsParams(
      @RequestParam Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }

  // 400 Bad Request, no params are retrieved therefore required params aren't met
  // this is analogous to the processFormPostAsParams solution, only it doesn't work with JSON
  @PostMapping(path="/json_with_params_skip_body", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostAsParams(
      @RequestParam Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
}
```

I was suspicious of Postman so I used telnet directly, with identical results.

All WWW Form requests have identical content (apart from the URI path):
```
POST /form_with_params HTTP/1.1
Host: localhost:8070
Accept: application/json
Content-Type: application/x-www-form-urlencoded
Content-Length: 30
Cache-Control: no-cache

client_id=foo&access_token=bar
```

Whereas all JSON request have this content:
```
POST /json_with_params HTTP/1.1
Host: localhost:8070
Accept: application/json
Content-Type: application/json
Content-Length: 40
Cache-Control: no-cache

{"client_id":"foo","access_token":"bar"}
```

**Comment by philwebb** [comment]
You can't treat elements from a JSON payload as request parameters, their not the same thing. Taking your examples:

```java
  // body is retrieved fine
  @PostMapping(path="/json",
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostBody(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
```

Here you're posting JSON, the `Jackson2JsonMessageConverter` is used and the JSON body is converted to a map and passed in.


```java
  // 400 Bad Request, no params are retrieved therefore required params aren't met
  @PostMapping(path="/json_with_params", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostBodyAndParams(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
```

Here you're posting JSON again but trying to restrict the `params`. The JSON payload doesn't get converted to request parameters so there are none. Hence the 400 response.


```java
  // 415 Unsupported media type
  // Content type 'application/x-www-form-urlencoded;charset=UTF-8' not supported
  @PostMapping(path="/form_with_params", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
  public ResponseEntity<String> processFormPostBodyAndParams(
      @RequestBody Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
```

Here you're posting form encoded data but then trying to read the body as a map. There's no converter for that so you get a 415.

```
  // body is retrieved fine, but only as long as it's injected as @RequestParam, not as @RequestBody
  @PostMapping(path="/form_with_params_skip_body", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
  public ResponseEntity<String> processFormPostAsParams(
      @RequestParam Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
```

Here things are working because `@RequestParam` can provide all request parameters. From the Javadoc:

> If the method parameter is Map<String, String> or MultiValueMap<String, String> and a parameter name is not specified, then the map parameter is populated with all request parameter names and values.


```java
  // 400 Bad Request, no params are retrieved therefore required params aren't met
  // this is analogous to the processFormPostAsParams solution, only it doesn't work with JSON
  @PostMapping(path="/json_with_params_skip_body", 
      params = {"client_id", "access_token"},
      consumes = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<String> processJsonPostAsParams(
      @RequestParam Map<String, String> body) {
    
    return new ResponseEntity<>("Body param map: " + body, HttpStatus.OK);
  }
}
```

This time you're asking for all request parameters, but there aren't any because the payload is JSON.

Basically, request parameters come from form encoded data and there is no mapping from JSON.

**Comment by lisandromc** [comment]
Phil, thanks for the quick reply.

If I understand correctly, specifying a RequestParam (be it from the params property or a @RequestParam annotation) causes the consumption of the payload (i.e., the request body), so that nothing is left by the time the `@RequestBody body` method param extraction is attempted. That would explain the necessity of the `@RequestParam body` fallback.

The issue to me was the fact that RequestParams could not be extracted from a JSON payload, unlike a URL-encoded payload. If RequestParams can only come from the URI query string or from a `x-www-form-urlencoded` payload, it might make sense to add a caveat to the [RequestParam doc page](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestParam.html) making that clear.

Thanks again for clearing this up.
Lisandro

**Comment by wilkinsona** [comment]
> it might make sense to add a caveat to the RequestParam doc page making that clear.

That documentation is part of Spring Framework rather than Spring Boot. Please open a [Spring Framework JIRA](https://jira.spring.io/browse/SPR) is you'd like to make your suggestion to the Framework team.
